### ISSUE TITLE
Generate type-safe accessor classes for gradle.properties

### ISSUE THREAD

**Issue by neyguvj** [gradle/gradle#19437]
in our project we often work with gradle.properties.
We created special accessor class to get type-safe access to gradle.properties.
It reads a map of project properties and assigns them to class fields.
For example we have a fileld in our class:
```
final class BuildParams {
    @Parameter(name = 'proguard.dont.obfuscate',
                         descriptionLines = ['Run proguard without obfuscation'])
    boolean dontObfuscate
}
```
In gradle.properties file we use property `proguard.dont.obfuscate=true`.
And in code we access to it with `BuildParams.dontObfuscate`.

i'we recently read article about [dependency catalog](https://docs.gradle.org/current/userguide/platforms.html) and i have an idea how to simplify working with properties.
We can declare them in settings.gradle like we declare libraries for dependency catalog.
```
projectParameters {
    progiard {
        parameter {
            name: 'disableObfuscation'
            description: 'Run proguard without obfuscation'
            type 'boolean'
            default 'false'
        }
    }
}
```
And then we can access to it in build.gradle files `Params.proguard.disableObfuscation`.

**Comment by ov7a** [comment]
Sorry for the late reply.

This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

**Comment by eskatos** [comment]
It seems to me that https://gradlex.org/build-parameters/ could be what you are looking for

**Comment by mlopatkin** [comment]
Having a first-class support of build parameters comes up from time to time, but it isn't something we're actively working on. I didn't find an existing issue so I filed #30504 to track further work here.

**Comment by vlsi** [comment]
I've been prototyping exactly this as a **Settings** plugin — type-safe accessors for build parameters, generated with ASM, no extra project and no `javac` — and I'd like to share what works, where it gets blocked, and a small API that would unblock it. Since this thread already mentions [gradlex build-parameters](https://gradlex.org/build-parameters/), it's worth spelling out why that plugin needs an extra Gradle build and why a Settings plugin can't avoid it today. The underlying limitation is tracked on the gradlex side as [gradlex-org/build-parameters#243](https://github.com/gradlex-org/build-parameters/issues/243).

**Why the current approach is costly.** The gradlex `build-parameters` plugin can't generate the accessors from `settings.gradle`; it generates Java sources into a separate included build and compiles them with `javac`. That has a real cost: every consumer build gains an extra Gradle build, and configuration is serialised, because the rest of the build waits for that included build to be configured and compiled before it can proceed, and the work can't be parallelised. For a feature whose only output is a handful of tiny accessor classes, that is a lot of overhead.

**What a Settings plugin can do today, and the blocker.** A Settings plugin can generate the accessor bytecode directly (ASM, no `javac`, no extra project) and resolve values lazily through `providers.gradleProperty(...)`, which is Configuration Cache and Isolated Projects safe. The missing piece is making the generated classes visible to build scripts (compile and runtime) and to the IDE. The only window to export them into the base project `ClassLoaderScope` is inside `BuildTreePreparingProjectsPreparer.prepareProjects`: the base scope is created, the version-catalog accessors are generated and exported into it, and then the scope is locked. All of this happens before `gradle.projectsLoaded` fires, so a Settings plugin has no hook in that window, and exporting at `projectsLoaded` lands on an already-locked, already-materialised scope and has no effect. It happens to work under TestKit's `withPluginClasspath()`, but not under a real `plugins { id(...) }` resolution.

**Gradle already has the machinery; it just isn't public.** Type-safe version-catalog accessors do all of this through the internal `DependenciesAccessors` service, via three seams:
- `generateAccessors(builders, baseScope, settings)`, called from `BuildTreePreparingProjectsPreparer` before the base scope is locked, with generation cached through the execution-engine workspace;
- `getClasses()` and `getSources()`, folded into the Kotlin DSL script model by `ProjectInternal.accessorsClassPathOf(...)` and served to IDEs by `KotlinBuildScriptModelBuilder` (this is what makes `libs.*` resolve in the editor);
- `createExtensions(project)`, called per project from `ProjectFactory`.

**Proposed minimal API.** Expose a small, incubating SPI that lets a Settings plugin contribute generated accessors through the same path. The plugin keeps ownership of its own DSL and code generation; Gradle owns export-before-lock, caching, and IDE surfacing:

```java
// Available to a Settings plugin; call during settings evaluation.
public interface GeneratedAccessors {
    void registerAccessors(GeneratedAccessorsProvider provider);
}

public interface GeneratedAccessorsProvider {
    // Invoked once, before the base project ClassLoaderScope is locked; the result is cached across builds.
    Result generate(Context context);

    interface Context {
        Directory getWorkspace();   // Gradle-managed, content-addressed, Configuration Cache safe
    }

    interface Result {
        ClassPath getClasses();     // exported to build scripts and run against
        ClassPath getSources();     // optional; surfaced to the IDE for navigation
    }
}
```

Per-project extension registration can stay the plugin's responsibility through the existing `gradle.lifecycle.beforeProject(IsolatedAction)`, which already works once the classes are on the scope, so it need not be part of the SPI.

This is the same mechanism #2885 asks for at the framework level, scoped down to a single extension point so plugins like `build-parameters` can deliver it without an extra build. I have a working spike (ASM generation, Configuration Cache and Isolated Projects green) and am happy to share it or help prototype the hook if there's interest.

**Comment by mlopatkin** [comment]
@vlsi thank you for the suggestion. Unfortunately, we don't have capacity to work on this right now, and that includes reviewing community contributions in the area. And because Declarative Gradle may influence the design or even feasibility of the feature significantly, we'd like to wait until certain decisions there are resolved.

**Comment by vlsi** [comment]
> And because Declarative Gradle may influence the design or even feasibility of the feature significantly, we'd like to wait until certain decisions there are resolved.

Frankly, "build flags" would still be required even with Declarative Gradle. Declarative Gradle does not resolve that use case.

**Comment by mlopatkin** [comment]
> > And because Declarative Gradle may influence the design or even feasibility of the feature significantly, we'd like to wait until certain decisions there are resolved.
> 
> Frankly, "build flags" would still be required even with Declarative Gradle. Declarative Gradle does not resolve that use case.

Definitely, but we may not need generated accessors (e.g. build flags become constants exported by plugin classes, or similar).

**Comment by vlsi** [comment]
> Definitely, but we may not need generated accessors (e.g. build flags become constants exported by plugin classes, or similar)

I am afraid that does not help build authors.
I do not think I understand your suggestion. Do you suggest everybody should create a publishable plugin for a build flag they need?

For instance, imagine I have `-Prelease` build flag to switch between `-SNAPSHOT` and release versions (by default build runs with `-SNAPSHOT`, and with `-Prelease` it produces release versions).
Do you suggest I should create and publish a plugin for it?

Note: creating the plugin as yet another Gradle module (`gradlex-org/build-parameters` approach) is a bad fit as it brings lot of overheads (kotlin compiler, module tasks, ...). That is exactly the reason I ask an API to register custom accessors without going through full gradle module machinery.
